cmake/Subdirectories.cmake

```diff
@@ -153,24 +160,112 @@ endfunction()
+                endif()
+
+                if(gem_path)
+                    set_property(GLOBAL PROPERTY "@GEMROOT:${gem_name}@" "${gem_path}")
+                endif()
+            endif()
+
             if (gem_path)
                 list(APPEND gem_dirs ${gem_path})
             elseif(NOT gem_optional)
-                list(JOIN registered_external_subdirs "\n" external_subdirs_formatted)
+                # Sort the list so it is easier to search visually
+                list(SORT registered_external_subdirs COMPARE NATURAL CASE INSENSITIVE ORDER ASCENDING)
+                # Indent the text to be easier to read. If the indent is removed CMake will add an
+                # additional newline automatically because "non-indented text is formatted in 
+                # line-wrapped paragraphs delimited by newlines"
+                list(JOIN registered_external_subdirs "\n  " external_subdirs_formatted)
                 message(SEND_ERROR "The gem \"${gem_name}\""
-                " could not be found in any gem.json from the following list of registered external subdirectories:\n"
-                "${external_subdirs_formatted}")
+                " could not be found in any gem.json from the following list of registered external subdirectories:"
+                "\n  ${external_subdirs_formatted}")
                 break()
             endif()
         endforeach()
     endif()
     set(${output_gem_dirs} ${gem_dirs} PARENT_SCOPE)
 endfunction()
 
+#! Use cmake.py to get a resolved list of gem names and paths for this object (engine or project)
+#! If dependencies are resolved successfully, save each gem's resolved path in a global property
+#! named "@GEMROOT:${gem_name}@"
+function(resolve_gem_dependencies object_type object_path)
+
+    # Avoid resolving dependencies for the same object type and path multiple times
+    get_property(resolved_dependencies GLOBAL PROPERTY "O3DE_RESOLVED_GEM_DEPENDENCIES_${object_type}_${object_path}")
+    if(resolved_dependencies)
+        return()
+    endif()
+
+    set(ENV{PYTHONNOUSERSITE} 1)
+    cmake_path(SET output_path "${CMAKE_BINARY_DIR}/${object_type}_external_subdirectories.out")
+    message(VERBOSE "Writing resolved gem dependencies for ${object_path} to '${output_path}'")
+    string(TOLOWER ${object_type} object_type_lower)
+    execute_process(COMMAND 
+        ${LY_PYTHON_CMD} "${LY_ROOT_FOLDER}/scripts/o3de/o3de/cmake.py" --${object_type_lower}-path "${object_path}" -gpof "${output_path}"
+        WORKING_DIRECTORY ${LY_ROOT_FOLDER}
+        RESULT_VARIABLE O3DE_CLI_RESULT
+        OUTPUT_VARIABLE O3DE_CLI_OUT 
+        ERROR_VARIABLE O3DE_CLI_OUT
+        )
+
+    if(O3DE_CLI_RESULT)
+        message(WARNING "Dependecy resolution failed\n  Error: ${O3DE_CLI_OUT}")
+        return()
+    endif()
+
+    file(READ "${output_path}" resolved_gem_dependency_output)
```

Comment:
Whoa, writing to a file during configuration this is a point of failure we should avoid.
This should write the dependencies to stdout and the OUTPUT_VARIABLE should be used.
Having to write and read a file on disk opens up issues with file permissions and sync issues.

Reply:
We can default to writing to [stdout](https://stackoverflow.com/a/7864158) if no output path is giving

Reply:
Sounds good, and I agree, it'd be safer. I went the file route to make it easier to troubleshoot failures and assumed if you can't write to the build folder then you've got bigger issues to solve 😄